applications/FluidDynamicsApplication/custom_processes/shock_detection_process.h

```diff
@@ -0,0 +1,366 @@
+     */
+    void ExecuteInitialize() override;
+
+    /**
+     * @brief Perform edge based shock detection
+     * This method performs the edge based shock detection
+     * @param rShockVariable Double variable to perform the shock detection
+     * @param rShockGradientVariable Vector variable to calculate the shock variable gradients
+     */
+    void EdgeBasedShockDetection(
+        const Variable<double>& rShockVariable,
+        const Variable<array_1d<double, 3>>& rShockGradientVariable);
+
+    /**
+     * @brief Perform edge based shock detection
+     * This method performs the edge based shock detection
+     * @param rShockVariable Component variable to perform the shock detection
+     * @param rShockGradientVariable Vector variable to calculate the shock variable gradients
+     */
+    void EdgeBasedShockDetection(
+        const VariableComponentType& rShockVariable,
+        const Variable<array_1d<double, 3>>& rShockGradientVariable);
+
+    /**
+     * @brief Template specialization of the edge based shock detection function
+     * Auxiliary method to specialize the variable types for the edge based shock detection
+     * @tparam TShockVariableType Shock variable type
+     * @tparam TShockGradientVariableType Shock gradient variable type
+     * @param rShockVariable Component variable to perform the shock detection
+     * @param rShockGradientVariable Vector variable to calculate the shock variable gradients
+     */
+    template<class TShockVariableType, class TShockGradientVariableType>
+    void EdgeBasedShockDetectionSpecialization(
+        const TShockVariableType& rShockVariable,
+        const TShockGradientVariableType& rShockGradientVariable)
+    {
+        // If required recompute the NODAL_AREA
+        // This is required for the nodal gradients calculation
+        if (mUpdateNodalAreaAtEachStep) {
+            CalculateNodalAreaProcess<CalculateNodalAreaSettings::SaveAsNonHistoricalVariable>(
+            mrModelPart,
+            mrModelPart.GetProcessInfo().GetValue(DOMAIN_SIZE)).Execute();
+        }
+
+        // If required recompute the NODAL_NEIGHBOURS
+        if (mUpdateNodalNeighboursAtEachStep) {
+            const auto& r_data_communicator = mrModelPart.GetCommunicator().GetDataCommunicator();
+            FindGlobalNodalNeighboursProcess(r_data_communicator, mrModelPart).Execute();
+        }
+
+        // Calculate the shock variable nodal gradients
+        ComputeNodalGradientProcess<ComputeNodalGradientProcessSettings::SaveAsNonHistoricalVariable>(
+            mrModelPart,
+            rShockVariable,
+            rShockGradientVariable).Execute();
+
+        auto& r_comm = mrModelPart.GetCommunicator();
+        auto& r_data_comm = r_comm.GetDataCommunicator();
+
+        // Create the global pointers list
+        GlobalPointersVector<Node<3>> global_pointers_list;
+        if (r_comm.IsDistributed()) {
+            for (auto &r_node : r_comm.LocalMesh().Nodes()) {
```

Comment:
ok, but you could have passed this as a lambda in the constructor.

Reply:
I think that in this way is more self-explained. If you prefer I can do it in a Lambda.